src/DetectChanges/BCBreak/ClassBased/EnumCasesChanged.php

```diff
@@ -0,0 +1,162 @@
+                if (self::isInternalDocComment($case->getDocComment())) {
+                    return false;
+                }
+
+                return ! $toClass->hasCase($case->getName());
+            },
+        );
+
+        $internalisedCases = array_filter(
+            $toClass->getCases(),
+            static function (ReflectionEnumCase $case) use ($fromClass) {
+                if (! self::isInternalDocComment($case->getDocComment())) {
+                    return false;
+                }
+
+                $fromClassCase = $fromClass->getCase($case->getName());
+                if (! $fromClassCase) {
+                    return false;
+                }
+
+                return ! self::isInternalDocComment($fromClassCase->getDocComment());
+            },
+        );
+
+        $nowNotInternalCases = array_filter(
+            $toClass->getCases(),
+            static function (ReflectionEnumCase $case) use ($fromClass) {
+                if (self::isInternalDocComment($case->getDocComment())) {
+                    return false;
+                }
+
+                $fromClassCase = $fromClass->getCase($case->getName());
+                if (! $fromClassCase) {
+                    return false;
+                }
+
+                return self::isInternalDocComment($fromClassCase->getDocComment());
+            },
+        );
+
+        $caseRemovedChanges = array_map(
+            static function (ReflectionEnumCase $case) use ($fromEnumName): Change {
+                $caseName = $case->getName();
+
+                return Change::removed('Case ' . $fromEnumName . '::' . $caseName . ' was removed');
+            },
+            $removedCases,
+        );
+
+        $caseAddedChanges = array_map(
+            static function (ReflectionEnumCase $case) use ($fromEnumName): Change {
+                $caseName = $case->getName();
+
+                return Change::added('Case ' . $fromEnumName . '::' . $caseName . ' was added');
+            },
+            $addedCases,
+        );
+
+        $caseBecameInternalChanges = array_map(
+            static function (ReflectionEnumCase $case) use ($fromEnumName): Change {
+                $caseName = $case->getName();
+
+                return Change::changed('Case ' . $fromEnumName . '::' . $caseName . ' was marked "@internal"');
+            },
+            $internalisedCases,
+        );
+
+        $caseBecameNotInternalChanges = array_map(
+            static function (ReflectionEnumCase $case) use ($fromEnumName): Change {
+                $caseName = $case->getName();
+
+                return Change::changed('Case ' . $fromEnumName . '::' . $caseName . ' had "@internal" removed');
+            },
+            $nowNotInternalCases,
+        );
+
+        return Changes::fromList(
+            ...$caseRemovedChanges,
+            ...$caseAddedChanges,
+            ...$caseBecameInternalChanges,
+            ...$caseBecameNotInternalChanges,
+        );
```

Comment:
Specifically, this allows configuration by downstream consumers:

`new EnumCasesChanged` vs `new EnumBecameNonEnum` vs `NonEnumBecameEnum` or such

Reply:
> Overall OK for this patch.
> 
> If you happen to touch this repo (in a different feature) again, though, I endorse splitting such various cases in a more detailed way, since you covered "everything enum" in one go, which is a bit much :-)

Yes, I did think this might have grown a bit much. It's partly because I started with the EnumBased as an interface etc, and didn't want to copy the logic for that sort of thing from ClassBased.

So also curious what you'd think about a refactor to reduce some of the duplicate / similar code between ClassBased, InterfaceBased and TraitBased

Reply:
> So also curious what you'd think about a refactor to reduce some of the duplicate / similar code between ClassBased, InterfaceBased and TraitBased

About this, I think "categorizing types of breakages" is more valuable than preventing duplication.

I did not work much on this repo in the last years, but my general idea is to improve the report with:

 * the code location of a **specific** breakage
 * a trace of a breakage: nested rules should be displayed, such as `non-final class -> method added` or `interface -> method added`
 * an explanation for every BC rule of why a specific rule is a breakage, so that users can be informed about reasoning, not just problems

That all leads to a lot of duplication that is still valuable, though :)

Reply:
> > So also curious what you'd think about a refactor to reduce some of the duplicate / similar code between ClassBased, InterfaceBased and TraitBased
> 
> About this, I think "categorizing types of breakages" is more valuable than preventing duplication.
> 
> I did not work much on this repo in the last years, but my general idea is to improve the report with:
> 
>     * the code location of a **specific** breakage
> 
>     * a trace of a breakage: nested rules should be displayed, such as `non-final class -> method added` or `interface -> method added`
> 
>     * an explanation for every BC rule of why a specific rule is a breakage, so that users can be informed about reasoning, not just problems
> 
> 
> That all leads to a lot of duplication that is still valuable, though :)

I'm not sure, I think you could still have all the categorisation without the code duplication. Maybe with an enum for the category!  I might try that refactor if/when I have another contribution to make.

Reply:
I'm fine with any refactoring that improves things, but be aware that besides some duplication, this library has little to no known problems :D